pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,662 @@
+	/// penalize it with the input parameter.
+	/// `interest_rate_with(0)` with returns the original interest_rate without any penalization
+	fn interest_rate_with(&self, penalty: T::Rate) -> Result<T::Rate, ArithmeticError> {
+		self.info
+			.interest_rate
+			.ensure_sub(self.written_off_status.penalty)?
+			.ensure_add(penalty)
+	}
+
+	fn last_updated_debt(&self) -> Result<T::Balance, DispatchError> {
+		if self.last_updated == T::Time::now().as_secs() {
+			T::InterestAccrual::current_debt(self.info.interest_rate, self.normalized_debt)
+		} else {
+			T::InterestAccrual::previous_debt(
+				self.info.interest_rate,
+				self.normalized_debt,
+				self.last_updated,
+			)
+		}
+	}
+
+	pub fn present_value(&self) -> Result<T::Balance, DispatchError> {
+		let debt = self.last_updated_debt()?;
+		let debt = self.written_off_status.write_down(debt)?;
+
+		match &self.info.valuation_method {
+			ValuationMethod::DiscountedCashFlows(dcf) => {
+				// If the loan is overdue, there are no future cash flows to discount,
+				// hence we use the outstanding debt as the value.
+				let maturity_date = self.info.schedule.maturity.date();
+				if self.last_updated > maturity_date {
+					return Ok(debt);
+				}
+
+				Ok(dcf.compute_present_value(
+					debt,
+					self.last_updated,
+					self.info.interest_rate,
+					self.origination_date,
+					maturity_date,
+				)?)
+			}
+			ValuationMethod::OutstandingDebt => Ok(debt),
+		}
+	}
+
+	fn max_borrow_amount(&self) -> Result<T::Balance, DispatchError> {
+		Ok(match self.info.restrictions.max_borrow_amount {
+			MaxBorrowAmount::UpToTotalBorrowed { advance_rate } => advance_rate
+				.ensure_mul_int(self.info.collateral_value)?
+				.saturating_sub(self.total_borrowed),
+			MaxBorrowAmount::UpToOutstandingDebt { advance_rate } => advance_rate
+				.ensure_mul_int(self.info.collateral_value)?
+				.saturating_sub(T::InterestAccrual::current_debt(
+					self.info.interest_rate,
+					self.normalized_debt,
+				)?),
+		})
+	}
+
+	fn ensure_can_borrow(&self, amount: T::Balance) -> DispatchResult {
+		ensure!(
+			amount <= self.max_borrow_amount()?,
+			Error::<T>::from(BorrowLoanError::MaxAmountExceeded)
+		);
+
+		ensure!(
+			self.info.schedule.maturity.date() > T::Time::now().as_secs(),
+			Error::<T>::from(BorrowLoanError::MaturityDatePassed)
+		);
+
+		match self.info.restrictions.borrows {
+			BorrowRestrictions::WrittenOff => {
+				ensure!(
+					self.written_off_status.is_none(),
+					Error::<T>::from(BorrowLoanError::WrittenOffRestriction)
+				)
+			}
+		}
+
+		Ok(())
+	}
+
+	fn ensure_can_repay(&self, amount: T::Balance) -> Result<T::Balance, DispatchError> {
+		let current_debt =
+			T::InterestAccrual::current_debt(self.info.interest_rate, self.normalized_debt)?;
+
+		match self.info.restrictions.repayments {
+			RepayRestrictions::None => (),
+		};
+
+		Ok(amount.min(current_debt))
+	}
+
+	fn ensure_can_write_off(
```

Comment:
I really like the `ensure_can_` pattern here, cleanly separates the validation logic from the business logic.

Reply:
Thanks! 🙌🏻  
Yeah, it helps to separate the immutable and mutable parts